### DISCUSSION TITLE
What is F#+'s equivalent of Scott Wlaschin's `plus` function? (i.e., concatenating `Error`  values during a validation pipeline for example)

### DISCUSSION THREAD

**Discussion by toraritte** [gmpl/FSharpPlus#576] [Q&A] [answered]
The [A functional approach to error handling (Railway oriented programming)](https://swlaschin.gitbooks.io/fsharpforfunandprofit/content/posts/recipe-part2.html) article is a great intro to monadic error handling, and when I learned about the F#+ library, I tried to find the equivalent functions in it. The problem is that the article is informal on purpose, and it is hard to connect the dots with almost zero background knowledge (e.g., just learned that `>=>` is called Kleisli composition).

The `plus` function in question:

```
type Result<'TSuccess,'TFailure> = 
    | Success of 'TSuccess
    | Failure of 'TFailure

let plus addSuccess addFailure switch1 switch2 x = 
    match (switch1 x),(switch2 x) with
    | Success s1,Success s2 -> Success (addSuccess s1 s2)
    | Failure f1,Success _  -> Failure f1
    | Success _ ,Failure f2 -> Failure f2
    | Failure f1,Failure f2 -> Failure (addFailure f1 f2)
```

I went through the operators in the [`Operators` module](https://fsprojects.github.io/FSharpPlus/reference/fsharpplus-operators.html), but descriptions are terse and unfamiliar with abstractions in the notation (e.g., functors, monoids, applicatives).

Thank you for this library and also for the help in advance!

**Comment by wallymathieu** [comment, marked as answer]
Main starting point is the [abstractions page](https://fsprojects.github.io/FSharpPlus/abstractions.html). The abstraction you are looking for is called an applicative. An example of applicative error handling is [done here](https://github.com/wallymathieu/validation-studies/blob/4b88a1082802d9ea805f054e7eacbbe10d4b4ba0/src/FsharpPlus/Library.fs) as well as the documentation on the [applicative page](https://fsprojects.github.io/FSharpPlus/abstraction-applicative.html).

Note the usage of `<!>` and `<*>` that you can find referenced in the above mentioned Operators module and in the documentation for applicatives.

**Comment by toraritte** [comment]
Thank you!

> Note the usage of <!> and <*> that you can find referenced in the above mentioned Operators module

I remember looking at them, but looks like I couldn't put 2 and 2 together. Thanks for the hint!
